json-format.md

```diff
@@ -129,34 +129,85 @@ OPTIONAL not omitted attributes MAY be represeted as a `null` JSON value.
+content, a JSON serializer MUST store a string representation of the data value,
+properly encoded according to the `datacontenttype`, in the `data` member of the
+JSON representation. An implementation MAY fail to serialize the event if it is
+unable to represent the runtime value as a properly encoded string.
+
+Out of this follows that the presence of the `data` and `data_base64` members is
+mutually exclusive in a JSON serialized CloudEvent.
+
+Furthermore, unlike attributes, for which value types are restricted to strings
+per the [type-system mapping](#22-type-system-mapping), the `data` member
+[JSON value][json-value] is unrestricted, and MAY contain any valid JSON if the
+`datacontenttype` declares the data to be JSON-formatted. In particular, the
+`data` member MAY have a value of `null`, representing an explicit `null`
+payload as distinct from the absence of the `data` member.
+
+#### 3.1.2. Payload Deserialization
 
 When a CloudEvents is deserialized from JSON, the presence of the `data_base64`
 member clearly indicates that the value is a Base64 encoded binary data, which
-the serializer MUST decode into a binary runtime data type. When a `data` member
-is present, it is decoded using the default JSON type mapping for the used
-runtime.
-
-Unlike attributes, for which value types are restricted to strings per the
-[type-system mapping](#22-type-system-mapping), the resulting `data` member
-[JSON value][json-value] is unrestricted, and MAY contain any valid JSON.
+the deserializer MUST decode into a binary runtime data type. The deserializer
+MAY further interpret this binary data according to the `datacontenttype`.
+
+When a `data` member is present, the decoding behavior is dependent on the value
+of the `datacontenttype` attribute. If the `datacontenttype` declares the `data`
+to contain JSON-formatted content (that is, its subtype is `json` or has a
+`+json` format extension), then the `data` member MUST be treated directly as a
+[JSON value][json-value] and decoded using an appropriate JSON type mapping for
+the runtime. Note: if the `data` member is a string, a JSON deserializer MUST
+interpret it directly as a [JSON String][json-string] value; it MUST NOT further
+deserialize the string as a JSON document.
+
+If the `datacontenttype` does not declare JSON-formatted data content, then the
+`data` member SHOULD be treated as an encoded content string. An implementation
+MAY fail to deserialize the event if the `data` member is not a string, or if it
+is unable to interpret the `data` with the `datacontenttype`.
+
+When a `data` member is present, if the `datacontenttype` attribute is absent, a
+JSON deserializer SHOULD proceed as if it were set to `application/json`, which
+declares the data to contain JSON-formatted content. Thus, it SHOULD treat the
+`data` member directly as a [JSON value][json-value] as specified above.
+Furthermore, if a JSON-formatted event with no `datacontenttype` attribute, is
```

Comment:
This means that for at least the C# implementation, the result of deserializing a JSON-formatted event with no `datacontenttype` to a `CloudEvent` object will result in an object with the `DataContentType` property set to `"application/json"` exactly as if it had been set explicitly. There's no easy way of saying "Well, it wasn't really set, but it was implicitly application/json, so use that if it's reserialized." My guess is that other SDKs may be in the same position. Note that this will happen even if it's reserialized *with a JSON event formatter* - the act of "deserialize/reserialize" won't be a no-op here. (There are other cases where it might not be a no-op too, so I don't think this is fatal.)

I'm okay with that - I just want to check that we're all on the same page.

Reply:
Yes, that's how the Ruby SDK would behave as well. It's possible an SDK could keep the "original datacontenttype" and "effective datacontenttype" independently in the event object, but I don't think that's strictly necessary. It is important that the event semantics (i.e. the effective datacontenttype) do not change when the event is reserialized, even to a different format. It is _not_ important that deserialize-reserialize to the same format is always a noop (although an SDK is free to try to do so anyway.) Do you think it's worth trying to state this for clarity? It seems like if we wanted to say something like that, it belongs in the main spec rather than in a format-specific document.

Reply:
I think it would be quite useful to be specified somewhere, but I don't mind where. I was mostly checking that I wouldn't end up with a completely different implementation to everyone else.